### DISCUSSION TITLE
Computed Relations slow at first request or after idle

### DISCUSSION THREAD

**Discussion by MHC2000** [PostgREST/postgrest#3313] [Q&A] [answered]
Hello,

I'm using Postgrest 12.0.2 with postgresql 13.
Im Adding some additional 1:many columns in a request by a computed reference.
When I'm calling the query for the first time it takes over a minute after I get a response.
If I trigger the query again directly or in the next 5 minutes the response is below 100 ms.
After trying several attempts to optimize I figured the added content by computed relations (dt_129_lt_2956_fk) is the part which is slowing it down. But I don't know why it's working after the first query as wished.
Does Postgrest recache the relations between tables after some time?

The server reacts on other queries always fast around 100 and 300 ms.
Also tried the foreign key, but the same problem.

- Full query: `dt_129?select=id,modified,created,software_project:cid_1499_data,lt_2956:dt_129_lt_2956_fk(*),modified,created&order=title.asc,id.asc&is_software_open=is.true&limit=50&offset=0`

- View 129 looks like that:
```
CREATE OR REPLACE VIEW api.dt_129
 AS
 SELECT mt.id,
    mt.is_active,
    mt.created,
    mt.modified,
    mt.description,
    mt.is_software_open,
    mt.software_project,
    mt.software_project_branch,
    mt.title,
    cid_1499_data.summary AS cid_1499_data
   FROM software.software_major_versions mt
     LEFT JOIN LATERAL ( SELECT jsonb_build_object('id', ot.id, 'title', ot.title, 'label', ot.title) AS summary
           FROM software.software_projects ot
          WHERE ot.id = mt.software_project) cid_1499_data ON true;

ALTER TABLE api.dt_129
    OWNER TO admin_users;
```

- That is the linked table view:
- 
```
CREATE OR REPLACE VIEW api.dt_129_lt_2956
 AS
 SELECT lt.id,
    lt.created,
    lt.modified,
    lt.is_active,
    lt.software_major_version,
    lt.swip_i_step2development_model_range,
    cid_1576_data.summary AS cid_1576_data
   FROM software.smv2swip_i_step lt
     LEFT JOIN LATERAL ( SELECT jsonb_build_object('id', ot.id, 'title', ot.title, 'label', ot.title) AS summary
           FROM master_data.swip_i_step2development_model_range ot
          WHERE ot.id = lt.swip_i_step2development_model_range) cid_1576_data ON true
  WHERE lt.is_active = true;
```

- The computed relation:
```

CREATE OR REPLACE FUNCTION api.dt_129_lt_2956_fk(
	mt api.dt_129)
    RETURNS SETOF api.dt_129_lt_2956 
    LANGUAGE 'sql'
    COST 100
    STABLE PARALLEL UNSAFE
    ROWS 1000

AS $BODY$
                SELECT lt.* 
                    FROM api.dt_129_lt_2956 lt
                    WHERE lt.software_major_version = mt.id
            
$BODY$;
```

Maybe somebody has an idea why the speed is at the beginning always very slow and then is for 10 minutes after the first request  ok.

Thanks!

Manfred

**Comment by MHC2000** [comment]
I figured, that postgrest is reconnecting every 30 minutes. Is that normal?

08/Mar/2024:08:00:12 +0100: Retrying listening for notifications on the pgrst channel..
08/Mar/2024:08:00:12 +0100: Starting PostgREST 12.0.2...
08/Mar/2024:08:00:12 +0100: Attempting to connect to the database...
08/Mar/2024:08:00:12 +0100: Connection successful
08/Mar/2024:08:00:12 +0100: Config reloaded
08/Mar/2024:08:00:12 +0100: Listening for notifications on the pgrst channel
08/Mar/2024:08:00:14 +0100: Schema cache loaded
08/Mar/2024:08:30:12 +0100: Retrying listening for notifications on the pgrst channel..
08/Mar/2024:08:30:12 +0100: Starting PostgREST 12.0.2...
08/Mar/2024:08:30:12 +0100: Attempting to connect to the database...
08/Mar/2024:08:30:12 +0100: Connection successful
08/Mar/2024:08:30:12 +0100: Config reloaded
08/Mar/2024:08:30:12 +0100: Listening for notifications on the pgrst channel
08/Mar/2024:08:30:15 +0100: Schema cache loaded

**Reply by steve-chavez** [reply]
> 08/Mar/2024:08:30:12 +0100: Retrying listening for notifications on the pgrst channel..

@MHC2000 Looks like something is killing the PostgREST LISTEN channel. You can get that connection with:

```sql
select * from pg_stat_activity where application_name ilike '%postgrest%' and query ilike '%listen%';
```

~~(I'll add a message on the logs to be more clear why the retrying is done)~~ **Edit**: below it can be seen that there's a log line that prints the error. Though it doesn't follow the log format, which is a separate issue (https://github.com/PostgREST/postgrest/issues/3320).

Perhaps you have [idle_session_timeout](https://www.postgresql.org/docs/16/runtime-config-client.html#GUC-IDLE-SESSION-TIMEOUT) set to 30 mins on the db?

**Reply by steve-chavez** [reply]
> Perhaps you have [idle_session_timeout](https://www.postgresql.org/docs/16/runtime-config-client.html#GUC-IDLE-SESSION-TIMEOUT) set to 30 mins on the db?

A bit strange you don't get clear log lines when the above is set though. I get:

```
FATAL:  terminating connection due to idle-session timeout
08/Mar/2024:10:34:00 -0500: Retrying listening for notifications on the pgrst channel..
```

And when manually terminating the channel (`select pg_terminate_backend(pid) from pg_stat_activity where application_name ilike '%postgrest%' and query ilike '%listen%';`):

```
FATAL:  terminating connection due to administrator command
08/Mar/2024:10:35:12 -0500: Retrying listening for notifications on the pgrst channel..
```

**Reply by MHC2000** [reply]
@steve-chavez thanks for the analysis.
When I try: `select pg_terminate_backend(pid) from pg_stat_activity where application_name ilike '%postgrest%' and query ilike '%listen%`';
I see the same entries in the log as you do:
```
FATAL:  terminating connection due to administrator command
08/Mar/2024:17:49:51 +0100: Retrying listening for notifications on the pgrst channel..
08/Mar/2024:17:49:51 +0100: Starting PostgREST 12.0.2...
08/Mar/2024:17:49:51 +0100: Attempting to connect to the database...
08/Mar/2024:17:49:51 +0100: Connection successful
08/Mar/2024:17:49:51 +0100: Config reloaded
08/Mar/2024:17:49:51 +0100: Listening for notifications on the pgrst channel
08/Mar/2024:17:49:54 +0100: Schema cache loaded
```
In the table pg_settings I don't see any entry for **idle_session_timeout**, so I gues it's not set..

**Reply by steve-chavez** [reply]
Hmm, something is killing the notification channel and then causing a schema cache reload on recovery. Anything interesting on the database logs related to the `pgrst` channel?

Are you using the `NOTIFY` mechanism to reload? If you do not, you could disable it https://postgrest.org/en/v12/references/configuration.html#db-channel-enabled.

**Reply by MHC2000** [reply]
Sadly I don't have superuser rights or access to the file system of the database. The database is hosted and can't change the rights.
So I guess I don't have any possibility to access the server logs?

Currently I'm using the notify mechanism after changes to my exposed views and functions.
But I will shut it down for test porpuses.

**Reply by MHC2000** [reply]
So I've deactivated the NOTIFY mechanism and postgrest doesn't reload the schema every 30 minutes.
Also the query which was working quite differently every now and then is now stable.

But I'm still curious why the NOTIFY was triggered for postgrest.
I do have a procedure which triggers the NOTIFY after I rebuild a view or function for my api.
But this procedure is not called periodically and I don't have any automations on this database which could call this procedure.

Is it possible that the LISTEN channel of postgrest is disconnected every 30 minutes due to a database setting or something else?

**Reply by steve-chavez** [reply]
> Is it possible that the LISTEN channel of postgrest is disconnected every 30 minutes due to a database setting or something else?

Yes, but that's dependent on your database provider.

> In the table pg_settings I don't see any entry for idle_session_timeout, so I gues it's not set..

Maybe query `pg_db_role_settings` as well and check an entry for the connection role there

**Reply by MHC2000** [reply]
> Yes, but that's dependent on your database provider.

It's within my company, but I haven't found any hints that there is something like that.
Other connections are lasting longer than 30 minutes and after turning the channel off, no reconnects of postgrest.

> Maybe query pg_db_role_settings as well and check an entry for the connection role there

Just one entry with: {search_path=public}.
Nothing like connection role.

I guess the only way to figure that out, is to have an error message which tells where it's comming from.

**Reply by steve-chavez** [reply, marked as answer]
> I guess the only way to figure that out, is to have an error message which tells where it's comming from.

Right. Trying to improve that on https://github.com/PostgREST/postgrest/pull/3323

**Reply by MHC2000** [reply]
Perfect, thanks a lot for the help!

**Reply by steve-chavez** [reply]
@MHC2000 There are binaries (at the bottom of the page) that include the above fix on:  https://github.com/PostgREST/postgrest/actions/runs/8241905872

Could you run the one that works on your platform and see what the logs show now?

**Reply by MHC2000** [reply]
This is what I got with the test version:

```
12/Mar/2024:09:16:59 +0100: Starting PostgREST 12.1...
12/Mar/2024:09:16:59 +0100: Attempting to connect to the database...
12/Mar/2024:09:16:59 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 1                              3.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubunt                              u1~20.04.2) 9.4.0, 64-bit
12/Mar/2024:09:16:59 +0100: Listening on port 3000
12/Mar/2024:09:16:59 +0100: Config reloaded
12/Mar/2024:09:16:59 +0100: Listening for notifications on the pgrst channel
12/Mar/2024:09:17:02 +0100: Schema cache queried in 387.8 milliseconds
12/Mar/2024:09:17:02 +0100: Schema cache loaded 880 Relations, 3190 Relationship                              s, 1378 Functions, 0 Domain Representations, 4 Media Type Handlers
12/Mar/2024:09:18:40 +0100: Schema cache loaded in 96471.0 milliseconds
12/Mar/2024:09:46:59 +0100: FatalError {fatalErrorMessage = "Error checking for PostgreSQL notifications"}. Retrying listening for notifications on the pgrst channel..
12/Mar/2024:09:46:59 +0100: Attempting to connect to the database...
12/Mar/2024:09:46:59 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 13.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
12/Mar/2024:09:46:59 +0100: Config reloaded
12/Mar/2024:09:46:59 +0100: Listening for notifications on the pgrst channel
12/Mar/2024:09:47:02 +0100: Schema cache queried in 457.2 milliseconds
12/Mar/2024:09:47:02 +0100: Schema cache loaded 880 Relations, 3190 Relationships, 1378 Functions, 0 Domain Representations, 4 Media Type Handlers
12/Mar/2024:09:48:11 +0100: Schema cache loaded in 67784.3 milliseconds

```

So I guess there is an error during listening.
But I don't get why postgrest is reconnectiong, because without the listening to pgrst the connection is stable and there wasn't any reconnecting the last days. Or is this the fail safe after an error with the listening, reconnecting?

**Reply by steve-chavez** [reply]
> But I don't get why postgrest is reconnectiong, because without the listening to pgrst the connection is stable and there wasn't any reconnecting the last days. Or is this the fail safe after an error with the listening, reconnecting?

It needs to reconnect + refresh cache because there might have been notifications sent when the LISTEN channel was down. 

The pool conneciton must also be reset otherwise we might run into https://github.com/PostgREST/postgrest/issues/2620 again.

https://github.com/diogob/hasql-notifications/issues/20 should provide more details on the error message.

**Reply by MHC2000** [reply]
okay understood, makes sense.
so this means we have to wait for further investigation, until https://github.com/diogob/hasql-notifications/issues/20 has been adressed?

**Reply by steve-chavez** [reply]
Yes, that's fixed now.

Please try the new version on https://github.com/PostgREST/postgrest/actions/runs/8267974977 and share the logs.

**Reply by MHC2000** [reply]
That's the new log:

```
13/Mar/2024:18:58:00 +0100: Starting PostgREST 12.1...
13/Mar/2024:18:58:00 +0100: Attempting to connect to the database...
13/Mar/2024:18:58:00 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 13.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
13/Mar/2024:18:58:00 +0100: Listening on port 3000
13/Mar/2024:18:58:00 +0100: Config reloaded
13/Mar/2024:18:58:00 +0100: Listening for notifications on the pgrst channel
13/Mar/2024:18:58:03 +0100: Schema cache queried in 386.2 milliseconds
13/Mar/2024:18:58:03 +0100: Schema cache loaded 876 Relations, 3178 Relationships, 1356 Functions, 0 Domain Representations, 4 Media Type Handlers
13/Mar/2024:18:59:01 +0100: Schema cache loaded in 57733.9 milliseconds
13/Mar/2024:19:28:00 +0100: Could not listen for notifications on the pgrst channel. SSL SYSCALL error: EOF detected Retrying listening for notifications..
13/Mar/2024:19:28:00 +0100: Attempting to connect to the database...
13/Mar/2024:19:28:00 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 13.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
13/Mar/2024:19:28:00 +0100: Config reloaded
13/Mar/2024:19:28:00 +0100: Listening for notifications on the pgrst channel
13/Mar/2024:19:28:03 +0100: Schema cache queried in 371.2 milliseconds
13/Mar/2024:19:28:03 +0100: Schema cache loaded 876 Relations, 3178 Relationships, 1356 Functions, 0 Domain Representations, 4 Media Type Handlers
13/Mar/2024:19:29:08 +0100: Schema cache loaded in 64228.6 milliseconds
```
Does the " SSL SYSCALL error: EOF detected" error suggest it's a connection error?

**Reply by wolfgangwalther** [reply]
Are you connecting to PostgreSQL via unix socket or TCP/IP? We might need some keepalive settings, if it's TCP/IP, for the listener connection.

**Reply by steve-chavez** [reply]
Yes but the error now suggests the OS might be killing the connection. 

- Is PostgREST running on a different machine than the database? If so, maybe a firewall might be finishing the connection.
- Maybe one of these OS settings https://www.ibm.com/docs/en/bcfsoz?topic=tcpip-timeout-behavior-linux-application-server could be causing the error.
- Some SO links ([ref1](https://dba.stackexchange.com/questions/175085/postgresql-ssl-syscall-error-eof-detected), [ref2](https://stackoverflow.com/questions/24130305/postgres-ssl-syscall-error-eof-detected-with-python-and-psycopg)) suggest that this could be an OOM error.

If you have a way to reproduce the error let me know and I'll see if I can do something on our end.

**Reply by MHC2000** [reply]
- Connection is via TCP/IP. Connection string looks like that: "postgres://user:password@host.com:1234/db_name?sslmode=require"
- Postgrest runs on a different machine than the database. Postgrest is running on a linux SLES 15 system.
- Sometimes queries over postgrest create OOM errors on the database, but that's on the productive system when some queries a not so good designed view which takes quite some time to query. But the listener error also happens on my test system where no one is doing anything besides me. The last logs where created while nothing was done. No queries, nothing on the database or the server.

So my best guess is the firewall. 
I will investigate if there is some firewall running on the server where postgrest is located (also a hosted system where I'm not completely fully in charge of all settings).

Thanks for the hints and help!

I'll keep you posted.

**Reply by wolfgangwalther** [reply]
Maybe you can add some of the [keepalive related settings](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-KEEPALIVES) to your connection string. This could work around the firewall issue.

**Reply by MHC2000** [reply]
@wolfgangwalther Sadly didn't help. Still getting the disconnection.

**Reply by MHC2000** [reply]
At the end it's a proxy timeout of the database which is by default 30 mins.
My database admins says they don't allow permanent connections only an increas to 100 minutes but always will disconnect after the given time.
So listening to the notify channel for reload is sadly no option in my setup.

Thanks again for the help and analysis!

**Reply by steve-chavez** [reply]
> At the end it's a proxy timeout of the database which is by default 30 mins.

@MHC2000 Do you know which setting that is? I would like to reproduce the error.

I was thinking if we could reconnect without reloading the cache + reset pool on those SSL errors.. but that wouldn't be reliable because again a notification might be sent around that time. So not a good option.

>  13/Mar/2024:18:58:03 +0100: Schema cache queried in 386.2 milliseconds
> 13/Mar/2024:18:58:03 +0100: Schema cache loaded 876 Relations, 3178 Relationships, 1356 Functions, 0 Domain Representations, 4 Media Type Handlers
> 13/Mar/2024:18:59:01 +0100: Schema cache loaded in 57733.9 milliseconds

Another thing that I've noticed is that loading the schema cache for your db takes considerable time (~57 seconds), which might be causing the slowdown as well. That seems related to https://github.com/PostgREST/postgrest/issues/3046.

> So listening to the notify channel for reload is sadly no option in my setup.

For now you could use SIGUSR1 https://postgrest.org/en/v12/references/schema_cache.html#schema-cache-reloading

**Reply by MHC2000** [reply]
> @MHC2000 Do you know which setting that is? I would like to reproduce the error.

I don't think it's a database setting, for me it looks like a proxy before the datatabase. They didn't get into too much details.
I'm still trying to clarify how the they consider to run an application with NOTIFY. Because I have another application which also relies on this mechanism and has the same problem.

> I was thinking if we could reconnect without reloading the cache + reset pool on those SSL errors.. but that wouldn't be reliable because again a notification might be sent around that time. So not a good option.

Maybe as optional parameter, that it shall ignore disconnects and not reload the schema on reconnect to notify. For me it would be a acceptable pay off to use the notify mechanism.

> For now you could use SIGUSR1 https://postgrest.org/en/v12/references/schema_cache.html#schema-cache-reloading

Will do, notify is deactivated and solved several problems in the performance

> Another thing that I've noticed is that loading the schema cache for your db takes considerable time (~57 seconds), which might be causing the slowdown as well. That seems related to https://github.com/PostgREST/postgrest/issues/3046.

We have a significant higher amount of views, relations and functions, looks like that is a related problem. I don't mind if the startup is a bit longer as long the performance afterwards is good.

**Reply by steve-chavez** [reply]
> We have a significant higher amount of views, relations and functions, looks like that is a related problem. I don't mind if the startup is a bit longer as long the performance afterwards is good.

Right, checking a fix for that in https://github.com/PostgREST/postgrest/issues/3327

**Reply by steve-chavez** [reply]
@MHC2000 Try the new binary on https://github.com/PostgREST/postgrest/actions/runs/8299658712. Reloads should still happen but the usual response time shouldn't be affected.

**Reply by MHC2000** [reply]
Not sure if I'm using it wrong, but getting 503 when I run the version:  https://github.com/PostgREST/postgrest/actions/runs/8299658712
Also don't see the info the schema cache is done.

Log:
```
18/Mar/2024:10:55:17 +0100: Listening on port 3000
18/Mar/2024:10:55:17 +0100: Config reloaded
18/Mar/2024:10:55:17 +0100: Listening for notifications on the pgrst channel
18/Mar/2024:10:55:19 +0100: Schema cache queried in 375.6 milliseconds
127.0.0.1 - qqzcms2 [18/Mar/2024:10:55:42 +0100] "GET /languages?select=*&order=label HTTP/1.1" 503 - "https://localhost:4200/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
127.0.0.1 - qqzcms2 [18/Mar/2024:10:55:42 +0100] "GET /task_status?select=*&order=title HTTP/1.1" 503 - "https://localhost:4200/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
```

**Reply by steve-chavez** [reply]
@MHC2000 This version doesn't wait for the schema cache to load. Those 503 should be gone after a while, when the schema cache finishes loading.

**Reply by MHC2000** [reply]
@steve-chavez Sorry wasn't aware of that. So I've let it run longer and after 2-3 minutes the system was working finde.
After 30 minutes no entry for loosing the connection or similar and also no reloading of the cache.
Only saw that the schema cache was queried twice while starting the system.

But, the chache doesn't reload if it's triggered over NOTIFY or at least there is no entry in the log.

```
18/Mar/2024:18:06:05 +0100: Starting PostgREST 12.1 (4a2c851)...
18/Mar/2024:18:06:05 +0100: Attempting to connect to the database...
18/Mar/2024:18:06:05 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 13.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
18/Mar/2024:18:06:05 +0100: Listening on port 3000
18/Mar/2024:18:06:05 +0100: Listening for notifications on the pgrst channel
18/Mar/2024:18:06:05 +0100: Config reloaded
18/Mar/2024:18:06:08 +0100: Schema cache queried in 546.9 milliseconds
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:35 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:44 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:44 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:46 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:46 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
127.0.0.1 - qqzcms2 [18/Mar/2024:18:07:53 +0100] "GET /rpc/datatable_get_index HTTP/1.1" 503 - "" "got (https://github.com/sindresorhus/got)"
18/Mar/2024:18:06:08 +0100: Schema cache loaded 876 Relations, 3178 Relationships, 1356 Functions, 0 Domain Representations, 4 Media Type Handlers
18/Mar/2024:18:08:09 +0100: Schema cache loaded in 116875.0 milliseconds
18/Mar/2024:18:08:09 +0100: Attempting to connect to the database...
18/Mar/2024:18:08:09 +0100: Successfully connected to PostgreSQL 13.14 (Ubuntu 13.14-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
18/Mar/2024:18:08:09 +0100: Config reloaded
18/Mar/2024:18:08:12 +0100: Schema cache queried in 640.6 milliseconds
18/Mar/2024:18:08:12 +0100: Schema cache loaded 876 Relations, 3178 Relationships, 1356 Functions, 0 Domain Representations, 4 Media Type Handlers
18/Mar/2024:18:10:12 +0100: Schema cache loaded in 117593.8 milliseconds
```

**Reply by steve-chavez** [reply]
> After 30 minutes no entry for loosing the connection or similar and also no reloading of the cache.

@MHC2000 That shouldn't have changed, the `SSL SYSCALL error: EOF detected` should remain and the reconnection should happen. The main advantage now is that you won't have the slowdown on response times. Are you not seeing those effects?

> But, the chache doesn't reload if it's triggered over NOTIFY or at least there is no entry in the log.

On the second:

> 18/Mar/2024:18:08:09 +0100: Attempting to connect to the database...

It does show that the `NOTIFY` reached the server. Though that should be clearer.. I'll work on a fix.

**Reply by MHC2000** [reply]
The second reload at 18/Mar/2024:18:08:12 +0100 wasn't triggered, I did nothing at that time. the 2 reloads did postgrest by itself on the startup. 
I let the system run until 19:00. 
At 18:40, I was expecting the reconnect, but nothing happened. there is no reconnect, no log entry. 
at 18:45, I triggered a reload, nothing happened as well., no log entry. 
There were no reactions after 18:10, but there should have been two more events until 19:00.
Postgrest didn't crash, I could call the api until then.

**Reply by steve-chavez** [reply]
New binaries at https://github.com/PostgREST/postgrest/actions/runs/8340919851

Reloading with NOTIFY should now be clear in the logs. Please try it.

**Reply by MHC2000** [reply]
@steve-chavez Sorry, looks like my comment of last week wasn't set. 
The result with the version https://github.com/PostgREST/postgrest/actions/runs/8340919851 was the same. Didn't see any reload notifications when I triggered by notify a new schema reload. Also didn't see the reconnect after 30 minutes. API was always reachable

**Reply by steve-chavez** [reply]
New NOTIFY logs should show as described on https://github.com/PostgREST/postgrest/pull/3340. Dunno what could be wrong as the new logs are tested. I guess otherwise you can wait for a stable release.

**Reply by MHC2000** [reply]
ok, will wait for the release and check again

**Comment by MHC2000** [comment]
I've used the newest version 12.20 but sadly still having the issue, that after the disconnect the loading times are very high (round about 2 minutes) for the first call.
So I had to disable the notify-channel again to get things working.

Overall the first start is very slow, if there are tables loaded together over joins (also 2 minutes). 
If the table has no connections then it loads fast, also at the first start.

Running the system with 1100 views and 400 functions.
Maybe we are too big for it.

Wouldn't mind to have a long loading time in the first place, but the reloading at the running system is a problem and prevents me to use the normal notify mechanism.
